pkg/k8s/watchers/pod.go

```diff
@@ -350,18 +350,98 @@ func genServiceMappings(pod *slim_corev1.Pod, podIPs []string) []loadbalancer.SV
 						},
 					},
 				}
-				bes = append(bes, be)
+				if be.L3n4Addr.IP.To4() != nil {
+					bes4 = append(bes4, be)
+				} else {
+					bes6 = append(bes6, be)
+				}
+			}
+
+			// When HostIP is explicitly set, then we need to expose *only*
+			// on this address but not via other addresses. When it's not set,
+			// then expose via all local addresses. Same when the user provides
+			// an unspecified address (0.0.0.0 / [::]).
+			feIP := net.ParseIP(p.HostIP)
+			if feIP != nil && !feIP.IsUnspecified() {
+				// Migrate the loopback address into a 0.0.0.0 / [::]
+				// surrogate, thus internal datapath handling can be
+				// streamlined. It's not exposed for traffic from outside.
+				if feIP.IsLoopback() {
+					if feIP.To4() != nil {
+						feIP = net.IPv4zero
+					} else {
+						feIP = net.IPv6zero
+					}
+				}
+				fe := loadbalancer.L3n4AddrID{
+					L3n4Addr: loadbalancer.L3n4Addr{
+						IP: feIP,
+						L4Addr: loadbalancer.L4Addr{
+							Protocol: proto,
+							Port:     uint16(p.HostPort),
+						},
+						Scope: loadbalancer.ScopeExternal,
+					},
+					ID: loadbalancer.ID(0),
+				}
+				if fe.L3n4Addr.IP.To4() != nil {
+					fes4 = append(fes4, fe)
+				} else {
+					fes6 = append(fes6, fe)
+				}
+			} else {
+				nodeAddrAll := [2][]net.IP{
+					k.K8sSvcCache.GetNodeAddressing().IPv4().LoadBalancerNodeAddresses(),
+					k.K8sSvcCache.GetNodeAddressing().IPv6().LoadBalancerNodeAddresses(),
+				}
+				for _, addrs := range nodeAddrAll {
+					for _, ip := range addrs {
+						fe := loadbalancer.L3n4AddrID{
+							L3n4Addr: loadbalancer.L3n4Addr{
+								IP: ip,
+								L4Addr: loadbalancer.L4Addr{
+									Protocol: proto,
+									Port:     uint16(p.HostPort),
+								},
+								Scope: loadbalancer.ScopeExternal,
+							},
+							ID: loadbalancer.ID(0),
+						}
+						if fe.L3n4Addr.IP.To4() != nil {
+							fes4 = append(fes4, fe)
+						} else {
+							fes6 = append(fes6, fe)
+						}
+					}
+				}
+			}
+
+			// We don't have the node name available here, but in any
+			// case in the BPF data path we drop any potential non-local
+			// backends anyway (which should never exist in the first
+			// place), hence we can just leave it at Cluster policy.
+			if option.Config.EnableIPv4 && len(bes4) > 0 {
+				for _, fe4 := range fes4 {
+					svcs = append(svcs,
+						loadbalancer.SVC{
+							Frontend:      fe4,
+							Backends:      bes4,
+							Type:          loadbalancer.SVCTypeHostPort,
+							TrafficPolicy: loadbalancer.SVCTrafficPolicyCluster,
+						})
+				}
+			}
+			if option.Config.EnableIPv6 && len(bes6) > 0 {
+				for _, fe6 := range fes6 {
+					svcs = append(svcs,
+						loadbalancer.SVC{
+							Frontend:      fe6,
+							Backends:      bes6,
+							Type:          loadbalancer.SVCTypeHostPort,
+							TrafficPolicy: loadbalancer.SVCTrafficPolicyCluster,
+						})
+				}
```

Comment:
This can been simplified as:
```
			// When HostIP is explicitly set, then we need to expose *only*
			// on this address but not via other addresses. When it's not set,
			// then expose via all local addresses. Same when the user provides
			// an unspecified address (0.0.0.0 / [::]).
			feIP := net.ParseIP(p.HostIP)
			var nodeAddrAll [][]net.IP
			if feIP != nil && !feIP.IsUnspecified() {
				// Migrate the loopback address into a 0.0.0.0 / [::]
				// surrogate, thus internal datapath handling can be
				// streamlined. It's not exposed for traffic from outside.
				if feIP.IsLoopback() {
					if feIP.To4() != nil {
						feIP = net.IPv4zero
					} else {
						feIP = net.IPv6zero
					}
				}
				nodeAddrAll = [][]net.IP{
					{feIP},
				}
			} else {
				nodeAddrAll = [][]net.IP{
					k.K8sSvcCache.GetNodeAddressing().IPv4().LoadBalancerNodeAddresses(),
					k.K8sSvcCache.GetNodeAddressing().IPv6().LoadBalancerNodeAddresses(),
				}
			}
			for _, addrs := range nodeAddrAll {
				for _, ip := range addrs {
					fe := loadbalancer.L3n4AddrID{
						L3n4Addr: loadbalancer.L3n4Addr{
							IP: ip,
							L4Addr: loadbalancer.L4Addr{
								Protocol: proto,
								Port:     uint16(p.HostPort),
							},
							Scope: loadbalancer.ScopeExternal,
						},
						ID: loadbalancer.ID(0),
					}

					// We don't have the node name available here, but in any
					// case in the BPF data path we drop any potential non-local
					// backends anyway (which should never exist in the first
					// place), hence we can just leave it at Cluster policy.
					if fe.L3n4Addr.IP.To4() != nil {
						if option.Config.EnableIPv4 && len(bes4) > 0 {
							svcs = append(svcs,
								loadbalancer.SVC{
									Frontend:      fe,
									Backends:      bes4,
									Type:          loadbalancer.SVCTypeHostPort,
									TrafficPolicy: loadbalancer.SVCTrafficPolicyCluster,
								})
						}
					} else {
						if option.Config.EnableIPv6 && len(bes6) > 0 {
							svcs = append(svcs,
								loadbalancer.SVC{
									Frontend:      fe,
									Backends:      bes6,
									Type:          loadbalancer.SVCTypeHostPort,
									TrafficPolicy: loadbalancer.SVCTrafficPolicyCluster,
								})
						}
					}
				}
			}
```